File: Projects/Offensive/Reverse shell framework/sim_server.c
/* sim_server.c - Simple Reverse Shell Simulator (C)
   Safe: does NOT execute system commands.
   Compile: gcc -o sim_server sim_server.c
   Usage: ./sim_server [port] [token]
   Example: ./sim_server 9000 labtoken
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <arpa/inet.h>
#include <sys/socket.h>

#define BACKLOG 5
#define BUF_SIZE 4096

// helpers: send length-prefixed line
int send_msg(int fd, const char *s) {
    uint32_t n = htonl((uint32_t)strlen(s));
    if (write(fd, &n, 4) != 4) return -1;
    if (write(fd, s, strlen(s)) != (ssize_t)strlen(s)) return -1;
    return 0;
}

char *recv_msg(int fd) {
    uint32_t n;
    ssize_t r = read(fd, &n, 4);
    if (r <= 0) return NULL;
    n = ntohl(n);
    if (n == 0) return strdup("");
    char *buf = malloc(n+1);
    size_t got = 0;
    while (got < n) {
        ssize_t rr = read(fd, buf+got, n-got);
        if (rr <= 0) { free(buf); return NULL;}
        got += rr;
    }
    buf[n] = 0;
    return buf;
}

// very small handler that returns canned outputs
void handle_command_simple(int client, const char *cmd) {
    if (strcmp(cmd, "whoami") == 0) {
        send_msg(client, "student");
    } else if (strcmp(cmd, "uname") == 0) {
        send_msg(client, "Linux lab-sim 5.10.0");
    } else if (strncmp(cmd, "echo ", 5) == 0) {
        send_msg(client, cmd + 5);
    } else if (strcmp(cmd, "date") == 0) {
        time_t t = time(NULL);
        char tmp[128];
        strftime(tmp, sizeof(tmp), "%Y-%m-%dT%H:%M:%SZ", gmtime(&t));
        send_msg(client, tmp);
    } else if (strcmp(cmd, "help") == 0) {
        send_msg(client, "Available: whoami, uname, echo, date, help");
    } else {
        send_msg(client, "Unknown command in simulator");
    }
}

int main(int argc, char **argv) {
    int port = (argc > 1) ? atoi(argv[1]) : 9000;
    const char *token = (argc > 2) ? argv[2] : "labtoken";
    int sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) { perror("socket"); return 1; }
    struct sockaddr_in sa = {0};
    sa.sin_family = AF_INET;
    sa.sin_addr.s_addr = htonl(INADDR_LOOPBACK); // bind to localhost by default
    sa.sin_port = htons(port);
    if (bind(sock, (struct sockaddr*)&sa, sizeof(sa)) < 0) { perror("bind"); return 1; }
    listen(sock, BACKLOG);
    printf("sim_server listening on 127.0.0.1:%d token='%s'\n", port, token);
    while (1) {
        int client = accept(sock, NULL, NULL);
        if (client < 0) { perror("accept"); continue; }
        char *auth = recv_msg(client);
        if (!auth) { close(client); continue; }
        if (strncmp(auth, "AUTH ", 5) != 0 || strcmp(auth + 5, token) != 0) {
            send_msg(client, "AUTH_FAIL");
            free(auth);
            close(client);
            continue;
        }
        send_msg(client, "AUTH_OK");
        send_msg(client, "NOTICE: Authenticated. SIMULATOR only.");
        free(auth);
        // interactive: receive commands
        while (1) {
            char *cmd = recv_msg(client);
            if (!cmd) break;
            if (strcmp(cmd, "exit") == 0) { free(cmd); break; }
            handle_command_simple(client, cmd);
            free(cmd);
        }
        close(client);
    }
    close(sock);
    return 0;
}

File: Projects/Offensive/Reverse shell framework/sim_server.py
#!/usr/bin/env python3
"""
sim_server.py - Reverse Shell Simulator (Server)
Safe: does NOT execute shell commands. Runs on localhost by default.
Usage: python3 sim_server.py [--host HOST] [--port PORT] [--token TOKEN] [--log LOGFILE]
Example: python3 sim_server.py --host 127.0.0.1 --port 9000 --token mysecret
"""
import argparse, socket, threading, json, struct, time, datetime, logging

# -- Configuration / safe defaults --
DEFAULT_HOST = "127.0.0.1"
DEFAULT_PORT = 9000
DEFAULT_TOKEN = "lab-token"
LOGFILE = "sim_server.log"
INACTIVITY_TIMEOUT = 300  # seconds

# -- Simple framed JSON protocol helpers (length-prefixed) --
def send_msg(conn, obj):
    data = json.dumps(obj).encode("utf-8")
    conn.sendall(struct.pack(">I", len(data)) + data)

def recv_msg(conn, timeout=10):
    conn.settimeout(timeout)
    hdr = b""
    try:
        while len(hdr) < 4:
            chunk = conn.recv(4 - len(hdr))
            if not chunk:
                return None
            hdr += chunk
    except socket.timeout:
        return None
    length = struct.unpack(">I", hdr)[0]
    data = b""
    while len(data) < length:
        chunk = conn.recv(length - len(data))
        if not chunk:
            return None
        data += chunk
    return json.loads(data.decode("utf-8"))

# -- Simulated command handlers (whitelisted) --
SIMULATED_CMDS = {
    "uname": lambda args: ("Linux lab-sim 5.10.0-0", "", 0),
    "whoami": lambda args: ("student", "", 0),
    "echo": lambda args: (" ".join(args), "", 0),
    "date": lambda args: (datetime.datetime.utcnow().isoformat() + "Z", "", 0),
    "help": lambda args: ("Available commands: " + ", ".join(sorted(SIMULATED_CMDS.keys())), "", 0)
}

def handle_command(cmd_line):
    parts = cmd_line.strip().split()
    if not parts:
        return ("", "empty command", 1)
    cmd = parts[0]
    args = parts[1:]
    if cmd in SIMULATED_CMDS:
        try:
            out, err, code = SIMULATED_CMDS[cmd](args)
            return (out, err, code)
        except Exception as e:
            return ("", f"handler error: {e}", 2)
    else:
        return ("", f"command '{cmd}' not recognized in simulator", 127)

# -- Per-connection worker --
def client_worker(conn, addr, token_expected, logger):
    logger.info(f"Connection from {addr}")
    # Authenticate
    auth = recv_msg(conn, timeout=10)
    if not auth or auth.get("type") != "auth" or auth.get("token") != token_expected:
        logger.warning(f"Auth failed from {addr}")
        send_msg(conn, {"type":"auth_result", "ok":False, "reason":"invalid token"})
        conn.close()
        return
    send_msg(conn, {"type":"auth_result", "ok":True})
    last_activity = time.time()
    send_msg(conn, {"type":"notice", "msg":"Authenticated. This is a SIMULATOR. No commands will be executed on the OS."})
    try:
        while True:
            msg = recv_msg(conn, timeout=INACTIVITY_TIMEOUT)
            if not msg:
                logger.info(f"Connection closed / timeout from {addr}")
                break
            last_activity = time.time()
            if msg.get("type") == "ping":
                send_msg(conn, {"type":"pong"})
                continue
            if msg.get("type") == "command":
                cmd_text = msg.get("command","")
                logger.info(f"CMD from {addr}: {cmd_text}")
                out, err, code = handle_command(cmd_text)
                resp = {"type":"result", "stdout":out, "stderr":err, "exit_code":code}
                send_msg(conn, resp)
                logger.info(f"RESP to {addr}: exit={code}")
            else:
                send_msg(conn, {"type":"error","msg":"unknown message type"})
    except Exception as e:
        logger.exception(f"Worker error for {addr}: {e}")
    finally:
        conn.close()
        logger.info(f"Disconnected {addr}")

def main():
    p = argparse.ArgumentParser()
    p.add_argument("--host", default=DEFAULT_HOST)
    p.add_argument("--port", type=int, default=DEFAULT_PORT)
    p.add_argument("--token", default=DEFAULT_TOKEN)
    p.add_argument("--log", default=LOGFILE)
    args = p.parse_args()

    logging.basicConfig(filename=args.log, level=logging.INFO, format="%(asctime)s %(levelname)s: %(message)s")
    logger = logging.getLogger("sim_server")
    logger.info("Starting Reverse Shell Simulator (server)")
    logger.info(f"Listening on {args.host}:{args.port}, token='{args.token}' (keep this secret for the lab)")

    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind((args.host, args.port))
    s.listen(5)
    try:
        while True:
            conn, addr = s.accept()
            t = threading.Thread(target=client_worker, args=(conn, addr, args.token, logger), daemon=True)
            t.start()
    except KeyboardInterrupt:
        logger.info("Shutting down")
    finally:
        s.close()

if __name__ == "__main__":
    main()
